akka-http-core/src/main/scala/akka/http/impl/engine/rendering/HttpResponseRendererFactory.scala

```diff
@@ -33,22 +33,22 @@ private[http] class HttpResponseRendererFactory(serverHeader: Option[headers.Ser
     }
 
   // as an optimization we cache the Date header of the last second here
-  @volatile private[this] var cachedDateHeader: (Long, Array[Byte]) = (0L, null)
-
+  private[this] var cachedDateHeader: (Long, Array[Byte]) = (0L, null)
   private def dateHeader: Array[Byte] = {
-    var (cachedSeconds, cachedBytes) = cachedDateHeader
-    val now = System.currentTimeMillis
-    if (now - 1000 > cachedSeconds) {
-      cachedSeconds = now / 1000
+    var cachedSeconds = cachedDateHeader._1
+    val now = currentTimeMillis()
+    val nowSeconds = now / 1000
+    if (nowSeconds > cachedSeconds) {
+      cachedSeconds = nowSeconds
       val r = new ByteArrayRendering(48)
-      dateTime(now).renderRfc1123DateTimeString(r ~~ headers.Date) ~~ CrLf
-      cachedBytes = r.get
-      cachedDateHeader = cachedSeconds -> cachedBytes
+      DateTime(now).renderRfc1123DateTimeString(r ~~ headers.Date) ~~ CrLf
+      cachedDateHeader = cachedSeconds -> r.get
     }
-    cachedBytes
+    cachedDateHeader._2
```

Comment:
"Unsafe" because seconds and the rendered header wouldn't necessarily correspond to another any more. But this may not be a problem since every update to the header is locally still an improvement over the previously cached header version. So the worst thing that could happen, is a) that one thread still sees and uses the version of the last second because it hasn't seen the update of the timestamp yet or b) that it sees the stale timestamp and redundantly updates the header value.